src/mpid/ch4/netmod/ucx/ucx_progress.c

```diff
@@ -6,21 +6,23 @@
 #include "mpidimpl.h"
 #include "ucx_impl.h"
 
-void *MPIDI_UCX_am_buf;         /* message buffer has global scope because ucx executes am_handler */
-
-void MPIDI_UCX_am_handler(void *request, ucs_status_t status, ucp_tag_recv_info_t * info)
+ucs_status_t MPIDI_UCX_am_handler(void *arg, void *data, size_t length, ucp_ep_h reply_ep,
+                                  unsigned flags)
 {
-    void *p_data;
-    void *in_data;
+    void *tmp, *p_data;
     size_t data_sz;
-    MPIDI_UCX_am_header_t *msg_hdr = (MPIDI_UCX_am_header_t *) MPIDI_UCX_am_buf;
 
-    p_data = in_data =
-        (char *) msg_hdr->payload + (info->length - msg_hdr->data_sz - sizeof(*msg_hdr));
+    /* need to copy the message data for alignment purposes */
+    tmp = MPL_malloc(length, MPL_MEM_BUFFER);
+    MPIR_Memcpy(tmp, data, length);
```

Comment:
So UCX also doesn't guarantee any minimum alignment?

Reply:
That's right. With the new API (>= 1.10.0) there is RNDV support for AMs so that we can save an extra copy. I think we can work on integrating that, but it is still quite new so this PR is probably a good starting point for now.